patches/server/0725-Improve-entity-damage-with-descriptive-return-type.patch

```diff
@@ -0,0 +1,1478 @@
+     }
+ 
+     private static void initIdleActivity(Brain<Zoglin> brain) {
+-        brain.addActivity(Activity.IDLE, 10, ImmutableList.of(new StartAttacking<>(Zoglin::findNearestValidAttackTarget), new RunSometimes(new SetEntityLookTarget(8.0F), UniformInt.of(30, 60)), new RunOne(ImmutableList.of(Pair.of(new RandomStroll(0.4F), 2), Pair.of(new SetWalkTargetFromLookTarget(0.4F, 3), 2), Pair.of(new DoNothing(30, 60), 1)))));
++        brain.addActivity(Activity.IDLE, 10, ImmutableList.of(new StartAttacking<>(Zoglin::findNearestValidAttackTarget), new RunSometimes<>(new SetEntityLookTarget(8.0F), UniformInt.of(30, 60)), new RunOne<>(ImmutableList.of(Pair.of(new RandomStroll(0.4F), 2), Pair.of(new SetWalkTargetFromLookTarget(0.4F, 3), 2), Pair.of(new DoNothing(30, 60), 1))))); // Paper - decompile error - generify behaviour constructors
+     }
+ 
+     private static void initFightActivity(Brain<Zoglin> brain) {
+-        brain.addActivityAndRemoveMemoryWhenStopped(Activity.FIGHT, 10, ImmutableList.of(new SetWalkTargetFromAttackTargetIfTargetOutOfReach(1.0F), new RunIf<>(Zoglin::isAdult, new MeleeAttack(40)), new RunIf<>(Zoglin::isBaby, new MeleeAttack(15)), new StopAttackingIfTargetInvalid()), MemoryModuleType.ATTACK_TARGET);
++        brain.addActivityAndRemoveMemoryWhenStopped(Activity.FIGHT, 10, ImmutableList.of(new SetWalkTargetFromAttackTargetIfTargetOutOfReach(1.0F), new RunIf<>(Zoglin::isAdult, new MeleeAttack(40)), new RunIf<>(Zoglin::isBaby, new MeleeAttack(15)), new StopAttackingIfTargetInvalid<>()), MemoryModuleType.ATTACK_TARGET); // Paper - decompile error - generify behaviour constructors
+     }
+ 
+     private Optional<? extends LivingEntity> findNearestValidAttackTarget() {
+@@ -158,12 +158,13 @@ public class Zoglin extends Monster implements Enemy, HoglinBase {
+         return (double)this.getBbHeight() - (this.isBaby() ? 0.2D : 0.15D);
+     }
+ 
++    // Paper start - improved entity damage return value
+     @Override
+-    public boolean hurt(DamageSource source, float amount) {
+-        boolean bl = super.hurt(source, amount);
++    public EntityHurtResult tryToHurt(DamageSource source, float amount) {
++        final EntityHurtResult bl = super.tryToHurt(source, amount);
+         if (this.level.isClientSide) {
+-            return false;
+-        } else if (bl && source.getEntity() instanceof LivingEntity) {
++            return EntityHurtResult.NOT_HURT;
++        } else if (bl.isHurt() && source.getEntity() instanceof LivingEntity) {
+             LivingEntity livingEntity = (LivingEntity)source.getEntity();
+             if (this.canAttack(livingEntity) && !BehaviorUtils.isOtherTargetMuchFurtherAwayThanCurrentAttackTarget(this, livingEntity, 4.0D)) {
+                 this.setAttackTarget(livingEntity);
+@@ -174,6 +175,7 @@ public class Zoglin extends Monster implements Enemy, HoglinBase {
+             return bl;
+         }
+     }
++    // Paper end
```

Comment:
This comment seems to be way out on its own. Should be right after the last Paper change.